Source/Core/Core/HW/EXI/EXI_Channel.cpp

```diff
@@ -238,48 +240,87 @@ void CEXIChannel::DoState(PointerWrap& p)
-    }
+    if (type != m_devices[device_index]->m_device_type)
+      AddDevice(EXIDevice_Create(type, m_channel_id, m_memcard_header_data), device_index, false);
+
+    m_devices[device_index]->DoState(p);
 
-    if (type == EXIDEVICE_MEMORYCARDFOLDER && old_header_data != m_memcard_header_data &&
-        !Movie::IsMovieActive())
+    bool is_memcard = type == EXIDEVICE_MEMORYCARD || type == EXIDEVICE_MEMORYCARDFOLDER;
+    if (is_memcard && !Movie::IsMovieActive())
     {
-      // We have loaded a savestate that has a GCI folder memcard that is different to the virtual
-      // card that is currently active. In order to prevent the game from recognizing this card as a
-      // 'different' memory card and preventing saving on it, we need to reinitialize the GCI folder
-      // card here with the loaded header data.
-      // We're intentionally calling ExpansionInterface::ChangeDevice() here instead of changing it
-      // directly so we don't switch immediately but after a delay, as if changed in the GUI. This
-      // should prevent games from assuming any stale data about the memory card, such as location
-      // of the individual save blocks, which may be different on the reinitialized card.
-      // Additionally, we immediately force the memory card to None so that any 'in-flight' writes
-      // (if someone managed to savestate while saving...) don't happen to hit the card.
-      // TODO: It might actually be enough to just switch to the card with the
-      // notify_presence_changed flag set to true? Not sure how software behaves if the previous and
-      // the new device type are identical in this case. I assume there is a reason we have this
-      // grace period when switching in the GUI.
-      AddDevice(EXIDEVICE_NONE, device_index);
-      ExpansionInterface::ChangeDevice(m_channel_id, EXIDEVICE_MEMORYCARDFOLDER, device_index,
-                                       CoreTiming::FromThread::CPU);
+      // We are processing a savestate that has a memory card plugged into the system, but don't
+      // want to actually store the memory card in the savestate, so loading older savestates
+      // doesn't confusingly revert in-game saves done since then.
+
+      // Handling this properly is somewhat complex. When loading a state, the game still needs to
+      // see the memory card device as it was (or at least close enough) when the state was made,
+      // but at the same time we need to tell the game that the data on the card may have been
+      // changed and it should not assume that the data (especially the file system blocks) it may
+      // or may not have read before is still valid.
+
+      // To accomplish this we do the following:
+
+      CEXIMemoryCard* card_device = static_cast<CEXIMemoryCard*>(m_devices[device_index].get());
+
+      if (p.GetMode() != PointerWrap::MODE_READ)
+      {
+        // In any of the non-modifying modes, we just read the file system blocks from the currently
```

Comment:
You say "non-modifying modes" down here, but the `if` uses `mode != READ` - this seems a bit confusing. Any thoughts on reversing the condition and rewriting the comment to reflect the same intent?

Reply:
`MODE_READ` means 'read from state stream, write to passed data'... but yes, I see what you mean. I'll think about this.

Reply:
Better?